src/persist/src/indexed/columnar.rs

```diff
@@ -13,22 +13,43 @@
 use std::fmt;
 use std::iter::FromIterator;
 
-use serde::{Deserialize, Serialize};
+use arrow2::buffer::{Buffer, MutableBuffer};
+use arrow2::types::Index;
+use ore::cast::CastFrom;
 
-/// A set of ((Key, Val), Time, Diff) records stored in a columnar representation.
+pub mod arrow;
+pub mod parquet;
+
+/// A set of ((Key, Val), Time, Diff) records stored in a columnar
+/// representation.
+///
+/// Note that the data are unsorted, and unconsolidated (so there may be
+/// multiple instances of the same ((Key, Val), Time), and some Diffs might be
+/// zero, or add up to zero).
 ///
-/// Note that the data are unsorted, and unconsolidated (so there may be multiple
-/// instances of the same ((Key, Val), Time), and some Diffs might be zero, or
-/// add up to zero).
-#[derive(Clone, Default, Eq, PartialEq, Serialize, Deserialize)]
+/// The i'th key's data is stored in
+/// `key_data[key_offsets[i]..key_offsets[i+1]]`. Similarly for val.
+///
+/// Invariants:
+/// - key_offsets.len() == len + 1
+/// - key_offsets are non-decreasing
+/// - Each key_offset is <= key_data.len()
+/// - key_offsets.last().unwrap() == key_data.len()
```

Comment:
nit: is it right that the first offset here and for val offsets should be 0? (sorry for giving these as individual comments I wanted to jot these down before I forgot)